### DISCUSSION TITLE
K.feature.laf_is_inside_image returns False when the laf is inside the image

### DISCUSSION THREAD

**Discussion by ersincine** [kornia/kornia#1749] [Questions] [answered]
Either I misunderstand something about local affine frames or there is a bug. The code below displays **tensor([[False]])** when the LAF is obviously inside the image.

```
img = np.zeros((100, 100), dtype=np.uint8)

scale = torch.Tensor([[[[90]]]])
xy = torch.Tensor([[[50, 50]]])
ori = torch.Tensor([[[0]]])

laf = K.feature.laf.laf_from_center_scale_ori(xy, scale, ori)

print(K.feature.laf_is_inside_image(laf, K.image_to_tensor(img, False)))
```

In order to make sure that the scales of the LAFs are the diameter of the circles (and not the radius), I have tried the code below:

`kornia_moons.feature.visualize_LAF(K.image_to_tensor(img, False), laf)`

![plot](https://user-images.githubusercontent.com/12049328/174503458-b4db040e-fb6d-44d6-933a-6288419b74ea.png)

It looks like the LAF is entirely inside the given image. Now, where do I make the mistake?

**Edit:** I investigated the issue further after posting the question. K.feature.laf_is_inside_image internally uses K.feature.laf_to_boundary_points. I think the problem is there.

```
img = np.zeros((100, 100), dtype=np.uint8)
scale = 90
x = 50
y = 50
ori = 0

scale = torch.Tensor([[[[scale]]]])
xy = torch.Tensor([[[x, y]]])
ori = torch.Tensor([[[ori]]])

laf = K.feature.laf.laf_from_center_scale_ori(xy, scale, ori)

num_pts = 12
xy = K.feature.laf_to_boundary_points(laf, num_pts)[0]
scale = torch.Tensor([[[[1]]] * num_pts])
ori = torch.Tensor([[[0]] * num_pts])
laf = K.feature.laf.laf_from_center_scale_ori(xy, scale, ori)

kornia_moons.feature.visualize_LAF(K.image_to_tensor(img, False), laf)
```

The boundary points are not inside the image as you can see below. To me it looks like there is a bug in K.feature.laf_to_boundary_points.

![myplot](https://user-images.githubusercontent.com/12049328/174504128-e16c45bf-3e7a-4d83-8d63-f52371ddf7d9.png)

**Comment by ersincine** [comment, marked as answer]
I think I found the problem. It is kornia_moons... I don't know why but kornia_moons multiplies the scale of an LAF by 0.5 before visualizing. Interesting... @ducha-aiki or somebody who knows about kornia_moons.feature.visualize_LAF: Can you please explain why this is the case? Note: Now I assume the scale corresponds to the radius (and not the diameter) of an LAF when it is a circle.

**Reply by ducha-aiki** [reply]
Yes, scale is the radius of the feature, not the diameter

**Reply by ersincine** [reply]
> Yes, scale is the radius of the feature, not the diameter

@ducha-aiki Thank you for the clarification. I believe, however, there is a discrepancy: **kornia_moons.feature.visualize_LAF fails to visualize these LAFs properly** as the circle in the first image has a radius of 45 (when the scale is 90). Isn't this a bug? (In kornia, scale is the radius but in kornia_moons (at least for kornia_moons.feature.visualize_LAF) the given scale is interpreted as the diameter.)
